### ISSUE TITLE
Using ESC on whole package verification

### ISSUE THREAD

**Issue by itlchriss** [OpenJML/OpenJML#783]
Hi David,

    I am working on using OpenJML on verifying some open source packages, such as apache commons-collections-4.4.
The following command is used:
```
openjml -esc -sourcepath [path-to-source-code-folder]
```

There are 17 errors and 2 warnings generated. Most of the errors state that the types are incompatible. 
Do you have any suggestions on this problem? Or, does the command I have used correct?

Enclosing the results:

```
./commons-collections4-4.4-src/src/main/java/org/apache/commons/collections4/list/CursorableLinkedList.java:283: error: incompatible types: WeakReference<T> cannot be converted to WeakReference<Cursor<E>>
        cursors.add(new WeakReference<>(cursor));
                    ^
  where T,E are type-variables:
    T extends Object declared in class WeakReference
    E extends Object declared in class CursorableLinkedList
./commons-collections4-4.4-src/src/main/java/org/apache/commons/collections4/map/AbstractReferenceMap.java:539: error: incompatible types: DefaultMapEntry<K#1,V#1> cannot be converted to @org.jmlspecs.annotation.NonNull Entry<K#2,V#2>
                list.add(new DefaultMapEntry<>(entry));
                         ^
  where K#1,V#1,K#2,V#2 are type-variables:
    K#1 extends Object declared in class DefaultMapEntry
    V#1 extends Object declared in class DefaultMapEntry
    K#2 extends Object declared in class ReferenceEntrySet
    V#2 extends Object declared in class ReferenceEntrySet
./commons-collections4-4.4-src/src/main/java/org/apache/commons/collections4/FluentIterable.java:95: error: incompatible types: inference variable T#1 has incompatible bounds
        return of(IteratorUtils.asIterable(new SingletonIterator<>(singleton, false)));
                 ^
    equality constraints: T#2,E#1
    lower bounds: E#2
  where T#1,T#2,E#1,E#2 are type-variables:
    T#1 extends Object declared in method <T#1>of(Iterable<T#1>)
    T#2 extends Object declared in method <T#2>of(T#2)
    E#1 extends Object declared in method <E#1>asIterable(Iterator<? extends E#1>)
    E#2 extends Object declared in class SingletonIterator
./commons-collections4-4.4-src/src/main/java/org/apache/commons/collections4/EnumerationUtils.java:78: error: incompatible types: inference variable E#1 has incompatible bounds
        return IteratorUtils.toList(new EnumerationIterator<>(enumeration));
                                   ^
    equality constraints: E#2
    lower bounds: E#3
  where E#1,E#2,E#3 are type-variables:
    E#1 extends Object declared in method <E#1>toList(Iterator<? extends E#1>)
    E#2 extends Object declared in method <E#2>toList(Enumeration<? extends E#2>)
    E#3 extends Object declared in class EnumerationIterator
./commons-collections4-4.4-src/src/main/java/org/apache/commons/collections4/sequence/SequencesComparator.java:267: error: no suitable method found for append(KeepCommand<T#1>)
                    script.append(new KeepCommand<>(sequence1.get(i)));
                          ^
    method EditScript.append(KeepCommand<T#2>) is not applicable
      (argument mismatch; KeepCommand<T#1> cannot be converted to KeepCommand<T#2>)
    method EditScript.append(InsertCommand<T#2>) is not applicable
      (argument mismatch; KeepCommand<T#1> cannot be converted to InsertCommand<T#2>)
    method EditScript.append(DeleteCommand<T#2>) is not applicable
      (argument mismatch; KeepCommand<T#1> cannot be converted to DeleteCommand<T#2>)
  where T#1,T#2 are type-variables:
    T#1 extends Object declared in class KeepCommand
    T#2 extends Object declared in class SequencesComparator
./commons-collections4-4.4-src/src/main/java/org/apache/commons/collections4/sequence/SequencesComparator.java:272: error: no suitable method found for append(DeleteCommand<T#1>)
                        script.append(new DeleteCommand<>(sequence1.get(i)));
                              ^
    method EditScript.append(KeepCommand<T#2>) is not applicable
      (argument mismatch; DeleteCommand<T#1> cannot be converted to KeepCommand<T#2>)
    method EditScript.append(InsertCommand<T#2>) is not applicable
      (argument mismatch; DeleteCommand<T#1> cannot be converted to InsertCommand<T#2>)
    method EditScript.append(DeleteCommand<T#2>) is not applicable
      (argument mismatch; DeleteCommand<T#1> cannot be converted to DeleteCommand<T#2>)
  where T#1,T#2 are type-variables:
    T#1 extends Object declared in class DeleteCommand
    T#2 extends Object declared in class SequencesComparator
./commons-collections4-4.4-src/src/main/java/org/apache/commons/collections4/sequence/SequencesComparator.java:275: error: no suitable method found for append(InsertCommand<T#1>)
                        script.append(new InsertCommand<>(sequence2.get(j)));
                              ^
    method EditScript.append(KeepCommand<T#2>) is not applicable
      (argument mismatch; InsertCommand<T#1> cannot be converted to KeepCommand<T#2>)
    method EditScript.append(InsertCommand<T#2>) is not applicable
      (argument mismatch; InsertCommand<T#1> cannot be converted to InsertCommand<T#2>)
    method EditScript.append(DeleteCommand<T#2>) is not applicable
      (argument mismatch; InsertCommand<T#1> cannot be converted to DeleteCommand<T#2>)
  where T#1,T#2 are type-variables:
    T#1 extends Object declared in class InsertCommand
    T#2 extends Object declared in class SequencesComparator
./commons-collections4-4.4-src/src/main/java/org/apache/commons/collections4/sequence/SequencesComparator.java:287: error: no suitable method found for append(KeepCommand<T#1>)
                script.append(new KeepCommand<>(sequence1.get(i)));
                      ^
    method EditScript.append(KeepCommand<T#2>) is not applicable
      (argument mismatch; KeepCommand<T#1> cannot be converted to KeepCommand<T#2>)
    method EditScript.append(InsertCommand<T#2>) is not applicable
      (argument mismatch; KeepCommand<T#1> cannot be converted to InsertCommand<T#2>)
    method EditScript.append(DeleteCommand<T#2>) is not applicable
      (argument mismatch; KeepCommand<T#1> cannot be converted to DeleteCommand<T#2>)
  where T#1,T#2 are type-variables:
    T#1 extends Object declared in class KeepCommand
    T#2 extends Object declared in class SequencesComparator
./commons-collections4-4.4-src/src/main/java/org/apache/commons/collections4/map/MultiValueMap.java:111: error: cannot infer type arguments for MultiValueMap<>
        return new MultiValueMap<>(map, new ReflectionFactory<>(collectionClass));
               ^
./commons-collections4-4.4-src/src/main/java/org/apache/commons/collections4/bidimap/DualTreeBidiMap.java:249: error: incompatible types: DualTreeBidiMap<K#1,V#1> cannot be converted to SortedMap<K#2,V#2>
            super(new DualTreeBidiMap<>(sm, bidi.reverseMap, bidi.inverseBidiMap));
                  ^
  where K#1,V#1,K#2,V#2 are type-variables:
    K#1 extends Object declared in class DualTreeBidiMap
    V#1 extends Object declared in class DualTreeBidiMap
    K#2 extends Object declared in class ViewMap
    V#2 extends Object declared in class ViewMap
./commons-collections4-4.4-src/src/main/java/org/apache/commons/collections4/list/SetUniqueList.java:333: error: incompatible types: inference variable E#1 has incompatible bounds
        return ListUtils.unmodifiableList(new SetUniqueList<>(superSubList, subSet));
                                         ^
    equality constraints: E#2
    lower bounds: E#2
  where E#1,E#2 are type-variables:
    E#1 extends Object declared in method <E#1>unmodifiableList(List<? extends E#1>)
    E#2 extends Object declared in class SetUniqueList
./commons-collections4-4.4-src/src/main/java/org/apache/commons/collections4/list/TreeList.java:508: error: incompatible types: AVLNode<E> cannot be converted to AVLNode<E>
                setLeft(new AVLNode<>(-1, obj, this, left), null);
                        ^
  where E is a type-variable:
    E extends Object declared in class AVLNode
./commons-collections4-4.4-src/src/main/java/org/apache/commons/collections4/list/TreeList.java:523: error: incompatible types: AVLNode<E> cannot be converted to AVLNode<E>
                setRight(new AVLNode<>(+1, obj, right, this), null);
                         ^
  where E is a type-variable:
    E extends Object declared in class AVLNode
./commons-collections4-4.4-src/src/main/java/org/apache/commons/collections4/map/MultiKeyMap.java:197: error: incompatible types: MultiKey<K#1> cannot be converted to MultiKey<? extends K#2>
        decorated().addMapping(index, hashCode, new MultiKey<>(key1, key2), value);
                                                ^
  where K#1,K#2 are type-variables:
    K#1 extends Object declared in class MultiKey
    K#2 extends Object declared in class MultiKeyMap
./commons-collections4-4.4-src/src/main/java/org/apache/commons/collections4/map/MultiKeyMap.java:329: error: incompatible types: MultiKey<K#1> cannot be converted to MultiKey<? extends K#2>
        decorated().addMapping(index, hashCode, new MultiKey<>(key1, key2, key3), value);
                                                ^
  where K#1,K#2 are type-variables:
    K#1 extends Object declared in class MultiKey
    K#2 extends Object declared in class MultiKeyMap
./commons-collections4-4.4-src/src/main/java/org/apache/commons/collections4/map/MultiKeyMap.java:471: error: incompatible types: MultiKey<K#1> cannot be converted to MultiKey<? extends K#2>
        decorated().addMapping(index, hashCode, new MultiKey<>(key1, key2, key3, key4), value);
                                                ^
  where K#1,K#2 are type-variables:
    K#1 extends Object declared in class MultiKey
    K#2 extends Object declared in class MultiKeyMap
./commons-collections4-4.4-src/src/main/java/org/apache/commons/collections4/map/MultiKeyMap.java:624: error: incompatible types: MultiKey<K#1> cannot be converted to MultiKey<? extends K#2>
        decorated().addMapping(index, hashCode, new MultiKey<>(key1, key2, key3, key4, key5), value);
                                                ^
  where K#1,K#2 are type-variables:
    K#1 extends Object declared in class MultiKey
    K#2 extends Object declared in class MultiKeyMap
/home/chrissleong/openjml-ubuntu-17-14/specs/java/nio/file/Path.jml:25: warning: A non-pure method is being called where it is not permitted: java.nio.file.Path.isAbsolute()
    //@   requires path.isAbsolute() && path.canonicalPath.length != 0;
                                  ^
/home/chrissleong/openjml-ubuntu-17-14/specs/java/nio/file/Path.jml:28: warning: A non-pure method is being called where it is not permitted: java.nio.file.Path.isAbsolute()
    //@   requires !path.isAbsolute() && path.canonicalPath.length != 0;
                                   ^
Note: ./commons-collections4-4.4-src/src/main/java/org/apache/commons/collections4/collection/IndexedCollection.java uses or overrides a deprecated API.
Note: Recompile with -Xlint:deprecation for details.
Note: Some input files use unchecked or unsafe operations.
Note: Recompile with -Xlint:unchecked for details.
Note: Some messages have been simplified; recompile with -Xdiags:verbose to get full output
17 errors
2 warnings
```

**Comment by davidcok** [comment]
The command you cite does not list any files to be verified, so I'm not sure of which command you are using.
Second, the specification of the Java library is by no means complete, so in doing a project like the one you are attempting, you will encounter many Java and apache classes that need specifications added.
As for the specific type errors, those are on the face of it Java errors, not OpenJML. I'd need a more specific and smaller repro to look into it further.

**Comment by itlchriss** [comment]
Yes. I have changed the command from using '--sourcepath' to '--dir'. I suspect that the incompatible types issues are caused by implicit type casting between template types such as Map and HashMap with Object and Object[] types as arguments. I found this issues across several other open source projects as well, for instance, slf4j.
In addition, I agree with on inspecting smaller repro. Therefore, let's focus on issue #784 , I have attached a very specific error on the apache commons-collections4.4. Hope I can help on improving the tool.
